File: smart-pos-frontend/src/lib/api-error.ts
export type ApiErrorDetail = {
  field?: string;
  message?: string;
};

export type ApiErrorEnvelope = {
  timestamp?: string;
  status?: number;
  code?: string;
  message?: string;
  path?: string;
  details?: ApiErrorDetail[];
};

export class ApiError extends Error {
  readonly status: number;
  readonly code: string;
  readonly details: ApiErrorDetail[];

  constructor(status: number, code: string, message: string, details: ApiErrorDetail[] = []) {
    super(message);
    this.name = "ApiError";
    this.status = status;
    this.code = code;
    this.details = details;
  }

  static fromEnvelope(status: number, envelope: ApiErrorEnvelope | null, fallbackMessage: string): ApiError {
    const code = envelope?.code ?? "UNKNOWN_ERROR";
    const message = envelope?.message ?? fallbackMessage;
    const details = envelope?.details ?? [];
    return new ApiError(envelope?.status ?? status, code, message, details);
  }
}

File: backend/src/main/java/com/smartpos/backend/common/error/ApiError.java
package com.smartpos.backend.common.error;

import com.fasterxml.jackson.annotation.JsonInclude;

import java.time.Instant;
import java.util.List;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiError(
        Instant timestamp,
        int status,
        String code,
        String message,
        String path,
        List<FieldViolation> details
) {
    public record FieldViolation(String field, String message) {}

    public static ApiError of(int status, String code, String message, String path) {
        return new ApiError(Instant.now(), status, code, message, path, null);
    }

    public static ApiError of(int status, String code, String message, String path, List<FieldViolation> details) {
        return new ApiError(Instant.now(), status, code, message, path, details);
    }
}
